mitodl/ocw-studio
# overhaul site pipeline definition

# What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/1905

# Description (What does it do?)
This PR re-writes `site-pipeline.yml` as a class called `SitePipelineDefinition` that extends the `ol-concourse` `Pipeline` class.

# How can this be tested?
 - Spin up `ocw-studio` and make sure you have the following configured:
   - Concourse
   - Theme assets pipeline upserted & run
   - Minio
   - Github
   - Concourse `fly` CLI
 - Enter a Django shell using `docker-compose exec web ./manage.py shell`
 - Paste in the following code, replacing `test-site` with the `name` property of the site you wish to test:
```python
from websites.models import *
from content_sync.pipelines.definitions.concourse.site_pipeline import (
    SitePipelineDefinitionConfig,
    SitePipelineDefinition,
)

site_name = "6-041sc-probabilistic-systems-analysis-and-applied-probability-fall-2013"
site = Website.objects.get(name=site_name)
pipeline_config = SitePipelineDefinitionConfig(
    site=site,
    pipeline_name="draft",
    instance_vars=f"?vars=%7B%22site%22%3A%20%22{site_name}%22%7D",
    site_content_branch="preview",
    static_api_url="https://draft.ocw.mit.edu",
    storage_bucket_name="ol-ocw-studio-app",
    artifacts_bucket="ol-eng-artifacts",
    web_bucket="ocw-content-draft",
    offline_bucket="ocw-content-offline-draft",
    resource_base_url="",
    ocw_studio_url="http://10.1.0.102:8043",
    ocw_hugo_themes_branch="main",
    ocw_hugo_projects_branch="main",
)
pipeline_definition = SitePipelineDefinition(config=pipeline_config)
f = open("site-pipeline-test.yml", "w")
f.write(pipeline_definition.json(indent=2, by_alias=True))
f.close()
```
 - After running the above, you should have a file called `site-pipeline-test.yml` at the root of your repo
 - Use the `fly` CLI to push up the pipeline definition to your locally running Concourse, again replacing `test-site` with the `name` of your site:
```bash
fly -t local set-pipeline -p draft --instance-var site=test-site -c site-pipeline-test.yml
```
 - Open the Concourse web UI at http://localhost:8080/ and log in
 - Search for the pipeline we just pushed up by typing the `name` of your site in the search bar and click on the pipeline in the `draft` group
 - Open the `ocw-studio` UI at http://localhost:8043/ in a new tab, pull up your site and publish the draft version there
 - Switch back over to the Concourse UI
 - The pipeline run should have started, watch it and ensure that it completes
 - Manipulate the arguments to the `SitePipelineDefinition` object in the shell code above to push up different versions of the pipeline, for example:
   - `live` version of the above pipeline:
     - `pipeline_name="live"`
     - `site_content_branch="release"`
     - Remove the `--buildDrafts` argument from the Hugo args
     - In the `fly` command, `-p live` instead of `-p draft`
   - `draft` version of `ocw-www`:
     - `site_name = "ocw-www"`
     - `is_root_website=True`
     - `base_url="/"`
     - `static_resources_subdirectory="/ocw-www/`
     - remove `--baseUrl` arguments from Hugo commands
     - `delete_flag=""`


# Additional Context
This process should test much of the functionality, but some of it should be tested in RC. The branch that this PR merges, `cg/overhaul-site-pipeline`, does not include the changes to `concourse.py` that would actually switch over to using `SitePipelineDefinition` with the `backpopulate_pipelines` management command. So, this PR simply adds the class and tests for it.

# Screenshots
![image](https://github.com/mitodl/ocw-studio/assets/12089658/33562f68-30d6-4f49-8e32-ca383549af7e)
